### PR TITLE
Adds support for Enums as implemented in PHP 8.1

### PR DESCRIPTION
**Pull request by cfroystad**
[RFC](https://wiki.php.net/rfc/enumerations)

This PR implements Enumerations for PHP as implemented in PHP 8.1

```php
enum Suit {
  case Hearts;
  case Diamonds;
  case Clubs;
  case Spades;
}
```

turns into:

```
(enum_declaration
    (name)
    (enum_declaration_list 
      (enum_case (name))
      (enum_case (name))
      (enum_case (name))
      (enum_case (name))
   )
)
```

An enum in PHP might have methods and implement interfaces:
```php
enum Suit implements Bar, Baz {
  case Hearts;
  case Diamonds;
  case Clubs;
  case Spades;

  // Fulfills the interface contract.
  public function color(): string {
    return match($this) {
      Suit::Hearts, Suit::Diamonds => 'Red',
      Suit::Clubs, Suit::Spades => 'Black',
    };
  }
}
```

turns into:

```
(enum_declaration
    (name)
    (class_interface_clause
      (name)
      (name)
    )
    (enum_declaration_list 
      (enum_case (name))
      (enum_case (name))
      (enum_case (name))
      (enum_case (name))

      (comment)
      
      (method_declaration
        (visibility_modifier)
        (name)
        (formal_parameters)
        (type_list (primitive_type))
        (compound_statement
          (return_statement
            (match_expression
              (parenthesized_expression
                (variable_name (name))
              )
              (match_block
                (match_conditional_expression
                  (match_condition_list
                    (class_constant_access_expression (name) (name))
                    (class_constant_access_expression (name) (name))
                  )
                  (string)
                )
                (match_conditional_expression
                  (match_condition_list
                    (class_constant_access_expression (name) (name))
                    (class_constant_access_expression (name) (name))
                  )
                  (string)
                )
              )
            )
          )
        )
      )
   )
)
```

This change brings us to 100% support for PHP 8.1 (minus pending and draft PRs) scheduled for release in november 🎉 

Checklist:
- [x] All tests pass in CI
- [x] There are sufficient tests for the new fix/feature
- [x] Grammar rules have not been renamed unless absolutely necessary (0 rules renamed)
- [x] The conflicts section hasn't grown too much (0 new conflicts)
- [x] The parser size hasn't grown too much (master: 2116, PR: 2552)

### CODE REVIEW

**Review by maxbrunsfeld — APPROVED** [review]

**Comment by cfroystad** [comment]
After looking a bit more into this, I realized that the change increased the parser line count from 111 476 to 135 526. I'm not sure what's causing this and if there's anything to be done about it.

I'll see if I find the time to investigate, but if you have any hunches I'd be happy to learn!

File: grammar.js [hunk 1]
```diff
@@ -242,6 +243,35 @@ module.exports = grammar({
...
+      field('body', $.enum_declaration_list)
+    )),
+
+    enum_declaration_list: $ => seq(
+      '{',
+      repeat($._enum_member_declartaion),
+      '}',
+    ),
+
+    _enum_member_declartaion: $ => choice(
```
Comment (maxbrunsfeld) [inline grammar.js:261]: Oh, there's a typo in this name.

File: grammar.js [hunk 2]
```diff
@@ -242,6 +243,35 @@ module.exports = grammar({
...
+      $.enum_case,
+      $.method_declaration,
+      $.use_declaration,
+    ),
+
+    enum_case: $ => seq(
+      optional(field('attributes', $.attribute_list)),
+      'case',
+      field('name', $.name),
+      optional(field('value', seq('=', choice($.string, $.integer)))),
```
Comment (maxbrunsfeld) [inline grammar.js:271]: The `value` field should probably only be on the string/integer, and not on the equals sign.

**Review by maxbrunsfeld — COMMENTED** [review]
Sorry, I noticed a couple minor issues. Could you take a look at these comments when  you get a chance?

**Comment by cfroystad** [comment]
Most certainly! Both of those are almost embarrassing, so thanks for identifying them 🙂

File: grammar.js (hunk 1 above)
Comment (cfroystad) [reply grammar.js:261]: Fixed in the _bugfixes_ branch

File: grammar.js (hunk 2 above)
Comment (cfroystad) [reply grammar.js:271]: Fixed in the _bugfixes_ branch

**Pull request merged**